Documentation/read/query.md

```diff
@@ -0,0 +1,175 @@
+To implement a query, you basically need to implement the interface `IQueryFor<>`.
+The generic type argument should point to a [read model](read_model.md).
+This is a marker interface and does not require you implement anything specific.
+Instead it is relying on a convention; its looking for a property called `Query`.
+This property needs a getter and can effectively return any type. The type that it
+returns is recognized by Dolittle and Dolittle will look for a [query provider](query_providers.md)
+for the type returned. By default you could be returning an `IQueryable<>` - something
+Dolittle has a [query provider](query_providers.md) for.
+
+```csharp
+using Dolittle.Queries;
+
+public class AllEmployees : IQueryFor<Employee>
+{
+    public IQueryable<Employee> Query { get {....} }
+}
+```
+
+
+The implementation of the query can be anything you want it to be. Use whatever underlying datastore and technique to get to the data. The only rule is that there
+must be a [query provider](../../Extending/Read/query_providers.md) for the return type.
+
+Dolittle has a very simple [repository for read models](read_model_repository.md) that can be used. You don't have to use it, as you decide entirely what you use internally - the query sits there as a contract and can also then be changed to accommodate a change in storage strategy.
+
+### Arguments
+
+With queries, sometimes you need to filter based on input. Any public property with
+a getter and a setter is considered an argument for the query.
+
+These can then be part of your query.
+
+```csharp
+using Dolittle.Read;
+
+public class EmployeesHiredAfter : IQueryFor<Employee>
+{
+    IReadModelRepositoryFor<Employee> _repository;
+
+    public EmployeesHiredAfter(IReadModelRepositoryFor<Employee> repository)
+    {
+        _repository = repository;
+    }
+
+    public DateTime HiredAfter { get; set; }
+
+    public IQueryable<Employee> Query => _repository.Query.Where(e => e.HiredDate >= HiredAfter);
+}
+```
+
+
+## Validation
+
+The arguments can have validators associated with them. All you need to do is to create something
+that inherits from `QueryValidationDescriptorFor<>` and point the generic argument to the
+query type.
+
+```csharp
+using Dolittle.Validation;
+using Dolittle.Read.Validation;
+
+public class EmployeeHiredAfterValidator : QueryValidationDescriptorFor<EmployeesHiredAfter>
+{
+    public EmployeeHiredAfterValidator()
+    {
+        // Require that hiring date is set after a certain date
+        ForArgument(e => e.HiredAfter).HasToBeGreaterThan(new DateTime(1985,1,1));
+    }
+}
+```
+
+## Paging
+
+Paging is not something you need to think about. This is what the [query provider](query_provider.md)
+deals with. Paging is a frontend concern and should not something you need to think about.
+In fact, paging is ideally a view concern, something decided as part of the interaction design.
+The client representations of queries has functionality related to paging.
```

Comment:
I disagree. Paging cannot really be solved in a general way, and should be implemented as proper arguments to the Query. It places large restrictions on the interaction to the underlying data-storage to support paging in general, and having paging-like arguments that make sense in the language of the domain is better anyway (i.e. on other things than page-size and page-number, like from/to dates or highest-priority-issues). To really have paging based on "pages" you'd need to "lock-in" the read-side when you send the first query to avoid repeats across pages or missing read-models because they fell between pages.